### PR TITLE
IFD label prediction on multilabel interface

### PR DESCRIPTION
**Pull request by vesteinn**

### CODE REVIEW

File: src/greynirseq/nicenlp/models/multilabel.py [hunk 1]
```diff
@@ -188,23 +191,27 @@ def encode(self, sentence):
...
-
-        tokens = self.encode(sentence)
-        word_mask = torch.tensor([word_start_dict[t] for t in tokens.tolist()])
-        word_mask[0] = 0
-        word_mask[-1] = 0
-        word_mask = word_mask.unsqueeze(0)
-        tokens = tokens.unsqueeze(0)
-        (cat_logits, attr_logits), _extra = self.model(tokens, features_only=True, word_mask=word_mask)
-
+    def predict_labels(self, sentences):
```
Comment (HaukurPall) [inline src/greynirseq/nicenlp/models/multilabel.py:194]: Það væri flott að bæta við type annotation.

File: src/greynirseq/nicenlp/models/multilabel.py [hunk 2]
```diff
@@ -188,23 +191,27 @@ def encode(self, sentence):
...
+        cat_logits, attr_logits, word_mask = self.get_logits(sentences)
         labels = self.task.logits_to_labels(cat_logits, attr_logits, word_mask)
-
         return labels
 
+    def predict_ifd_labels(self, sentences):
```
Comment (HaukurPall) [inline src/greynirseq/nicenlp/models/multilabel.py:199]: Það væri flott að bæta við type annotation.

File: src/greynirseq/nicenlp/models/multilabel.py [hunk 3]
```diff
@@ -188,23 +191,27 @@ def encode(self, sentence):
...
+        labdict = self.model.task.label_dictionary
+        labels = self.predict_labels(sentences)
+        ifd_labels_batch = []
+        for sentence_labels in labels:
+            ifd_labels = []
+            for labelset in sentence_labels:
+                cat, feats = labelset
+                idxs = [labdict.symbols.index(label) for label in [cat] + feats]
+                oh = nn.functional.one_hot(torch.tensor(idxs), num_classes=len(labdict.symbols)).sum(dim=0)
+                oh = oh[labdict.nspecial + 1 :]
```
Comment (HaukurPall) [inline src/greynirseq/nicenlp/models/multilabel.py:210]: Af hverju `+1`?

**Review by HaukurPall — CHANGES_REQUESTED** [review]
Má bæta við annotations á public aðferðirnar og spurning með eitt `+ 1`

File: src/greynirseq/nicenlp/models/multilabel.py (hunk 3 above)
Comment (vesteinn) [reply src/greynirseq/nicenlp/models/multilabel.py:214]: See comment

File: src/greynirseq/nicenlp/models/multilabel.py [hunk 4]
```diff
@@ -177,27 +178,29 @@ def prepare_batch(self, sentences):
...
 
-    def encode(self, sentence):
+    def encode(self, sentence: str) -> str:
         # We add a space to treat word encoding the same at the front and back of a sentence.
         if sentence[0] != " ":
             sentence = " " + sentence
         return super().encode(sentence)
 
-    def decode(self, tokens, no_cut_space=False):
+    def decode(self, tokens: List[str], no_cut_space: bool = False) -> List[str]:
```
Comment (HaukurPall) [inline src/greynirseq/nicenlp/models/multilabel.py:194]: `no_cut_space` er ekki notað.

**Review by HaukurPall — APPROVED** [review]
